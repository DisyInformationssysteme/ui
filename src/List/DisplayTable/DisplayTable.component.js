import React, { PropTypes } from 'react';
import classnames from 'classnames';
import { Button } from 'react-bootstrap';
import Actions from '../../Actions';

import theme from './DisplayTable.scss';

function RowRenderer(props) {
	const { titleKey, onTitleClick, item } = props;
	return (
		<tr>
			{props.columns.map((column, index) => {
				if (column.key === titleKey) {
					const onClick = event => onTitleClick(
						event,
						item,
					);
					return (
						<td key={index}>
							<Button
								bsStyle="link"
								onClick={onClick}
								role="link"
							>
								{item[column.key]}
							</Button>
							<Actions
								actions={item.actions || []}
								hideLabel
								link
							/>
						</td>
					);
				}
				return (<td key={index}>{item[column.key]}</td>);
			})}
		</tr>
	);
}
RowRenderer.propTypes = {
	item: PropTypes.object, // eslint-disable-line react/forbid-prop-types
	columns: PropTypes.arrayOf(
		PropTypes.shape({
			key: PropTypes.string.isRequired,
		})
	).isRequired,
	titleKey: PropTypes.string,
	onTitleClick: PropTypes.func,
};

function ListHeader(props) {
	return (
		<tr>
			{props.columns.map((column, index) => (<th key={index}>{column.label}</th>))}
		</tr>
	);
}
ListHeader.propTypes = {
	columns: PropTypes.arrayOf(
		PropTypes.shape(
			{ label: PropTypes.string },
		)
	),
};


/**
 * @param {object} props react props
 * @example
<DisplayTable name="Hello world"></DisplayTable>
 */
function DisplayTable({ items, columns, titleKey, onTitleClick }) {
	const className = classnames(
		'table',
		'tc-list-display-table',
		theme.table,
	);
	return (
		<table className={className}>
			<thead>
				<ListHeader columns={columns} />
			</thead>
			<tbody>
				{items.map(
					(item, index) => (
						<RowRenderer
							key={index}
							item={item}
							columns={columns}
							titleKey={titleKey || 'name'}
							onTitleClick={onTitleClick}
						/>
					)
				)}
			</tbody>
		</table>
	);
}

DisplayTable.propTypes = {
	items: PropTypes.arrayOf(
		PropTypes.object
	),
	columns: PropTypes.arrayOf(
		PropTypes.object
	),
	titleKey: PropTypes.string,
	onTitleClick: PropTypes.func,
};

export default DisplayTable;
