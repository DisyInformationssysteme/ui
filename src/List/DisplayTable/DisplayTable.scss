// DisplayTable styles
.table :global(.btn-group) {
	visibility: hidden;
}

.table tr:hover :global(.btn-group) {
	visibility: visible;
}
