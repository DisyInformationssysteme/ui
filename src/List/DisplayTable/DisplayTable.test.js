import React from 'react';
import renderer from 'react-test-renderer';

import DisplayTable from './DisplayTable.component';

jest.mock('react-dom');

describe('DisplayTable', () => {
	it('should render its name', () => {
		const items = [
			{ id: 1, name: 'one' },
			{ id: 2, name: 'two' },
		];
		const columns = [
			{ key: 'id', label: 'Id' },
			{ key: 'name', label: 'Name' },
		];
		const onTitleClick = jest.fn();
		const wrapper = renderer.create(
			<DisplayTable
				items={items}
				columns={columns}
				onTitleClick={onTitleClick}
			/>).toJSON();
		expect(wrapper).toMatchSnapshot();
	});
});
