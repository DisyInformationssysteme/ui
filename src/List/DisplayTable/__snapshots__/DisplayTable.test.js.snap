exports[`DisplayTable should render with default title property (name) 1`] = `
<table
  className="table tc-list-display-table">
  <thead>
    <tr>
      <th>
        Id
      </th>
      <th>
        Name
      </th>
      <th>
        Author
      </th>
      <th>
        Created
      </th>
      <th>
        Modified
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
      </td>
      <td>
        <div
          style={
            Object {
              "display": "inline",
            }
          }>
          <span
            className={undefined}>
            Hello world
          </span>
        </div>
        <div
          className="tc-actions btn-group">
          <button
            aria-describedby={42}
            className="btn btn-link"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            role="link"
            type="button">
            <i
              aria-hidden="true"
              className="fa fa-edit"
              title={null} />
          </button>
          <button
            aria-describedby={42}
            className="btn btn-link"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            role="link"
            type="button">
            <i
              aria-hidden="true"
              className="fa fa-trash-o"
              title={null} />
          </button>
        </div>
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        <div
          style={
            Object {
              "display": "inline",
            }
          }>
          <span
            className={undefined}>
            Foo
          </span>
        </div>
        <div
          className="tc-actions btn-group" />
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        <div
          style={
            Object {
              "display": "inline",
            }
          }>
          <span
            className={undefined}>
            Bar
          </span>
        </div>
        <div
          className="tc-actions btn-group" />
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`DisplayTable should render with defined title property 1`] = `
<table
  className="table tc-list-display-table">
  <thead>
    <tr>
      <th>
        Id
      </th>
      <th>
        Name
      </th>
      <th>
        Author
      </th>
      <th>
        Created
      </th>
      <th>
        Modified
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <div
          style={
            Object {
              "display": "inline",
            }
          }>
          <i
            aria-hidden="true"
            className="fa fa-file-excel-o"
            title={null} />
          <button
            className="btn btn-link"
            disabled={false}
            onClick={[Function]}
            role="link"
            type="button">
            1
          </button>
        </div>
        <div
          className="tc-actions btn-group">
          <button
            aria-describedby={42}
            className="btn btn-link"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            role="link"
            type="button">
            <i
              aria-hidden="true"
              className="fa fa-edit"
              title={null} />
          </button>
          <button
            aria-describedby={42}
            className="btn btn-link"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            role="link"
            type="button">
            <i
              aria-hidden="true"
              className="fa fa-trash-o"
              title={null} />
          </button>
        </div>
      </td>
      <td>
        Hello world
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
    <tr>
      <td>
        <div
          style={
            Object {
              "display": "inline",
            }
          }>
          <i
            aria-hidden="true"
            className="fa fa-file-pdf-o"
            title={null} />
          <button
            className="btn btn-link"
            disabled={false}
            onClick={[Function]}
            role="link"
            type="button">
            2
          </button>
        </div>
        <div
          className="tc-actions btn-group" />
      </td>
      <td>
        Foo
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
    <tr>
      <td>
        <div
          style={
            Object {
              "display": "inline",
            }
          }>
          <button
            className="btn btn-link"
            disabled={false}
            onClick={[Function]}
            role="link"
            type="button">
            2
          </button>
        </div>
        <div
          className="tc-actions btn-group" />
      </td>
      <td>
        Bar
      </td>
      <td>
        Jean-Pierre DUPONT
      </td>
      <td>
        2016-09-22
      </td>
      <td>
        2016-09-22
      </td>
    </tr>
  </tbody>
</table>
`;
